csrc/sycl/ops.cpp

```diff
@@ -0,0 +1,2248 @@
+          [=](sycl::nd_item<3> item_ct1) {
+            kEstimateQuantiles<T>(buff_A, code, offset, std_numeric_limits_T_max_ct3, n, item_ct1, tacc);
+          });
+      });
+  }
+  //back memcpy
+  q_ct1.memcpy((T*)(A), (T*)(buff_A), NUM_BLOCK); 
+  
+}
+
+void quantize(float *code, float *A, unsigned char *out, int n)
+{
+  int num_blocks = n/1024;
+  num_blocks = n % 1024 == 0 ? num_blocks : num_blocks + 1;
+  dpct::device_ext &dev_ct1 = dpct::get_current_device();
+  sycl::queue &q_ct1 = dev_ct1.in_order_queue();
+  sycl::context ctx = q_ct1.get_context();
+  int size = NUM_BLOCK;
+  
+  *((float **)&buff_A) = sycl::malloc_device(size, A, ctx);
+  *((unsigned char **)&buff_out = sycl::malloc_device(size, out, ctx);
+  q_ct1.memcpy((float*)(buff_A), (float*)(A), NUM_BLOCK);
+  q_ct1.memcpy((unsigned char*)(buff_out), (unsigned char*)(out), NUM_BLOCK);
+
+  /*
+  DPCT1049:55: The work-group size passed to the SYCL kernel may exceed the limit. To get the device limit, query info::device::max_work_group_size. Adjust the work-group size if needed.
+  */
+  {
+    dpct::has_capability_or_fail(q_ct1.get_device(), {sycl::aspect::fp16});
+    q_ct1.submit(
+      [&](sycl::handler &cgh) {
+      using group_load = dpct::group::workgroup_load<NUM_BLOCK,BLOCK_LOAD_DIRECT,float>;
+      size_t load_temp_storage_size = group_load::get_local_memory_size(NUM_BLOCK);
+      using group_store = dpct::group::workgroup_store<NUM_BLOCK, BLOCK_STORE_DIRECT,unsigned char>;
+      size_t store_temp_storage_size = group_store::get_local_memory_size(NUM_BLOCK);
+      
+      sycl::local_accessor<uint8_t, 1> ltacc(load_temp_storage_size, cgh);
+      sycl::local_accessor<uint8_t, 1> stacc(store_temp_storage_size, cgh);
+      
+      //__shared__ vars
+      sycl::local_accessor<float, 1> smem_code_acc_ct1(sycl::range<1>(256), cgh);
+      
+      cgh.parallel_for(
+        sycl::nd_range<3>(sycl::range<3>(1, 1, num_blocks) * sycl::range<3>(1, 1, 1024), sycl::range<3>(1, 1, 1024)), 
+        [=](sycl::nd_item<3> item_ct1) {
+          kQuantize(code, buff_A, buff_out, n, item_ct1, smem_code_acc_ct1.get_pointer(), ltacc, stacc);
+        });
+    });
+  }
+  //back memcpy
+  q_ct1.memcpy((float*)(A), (float*)(buff_A), NUM_BLOCK);
+  q_ct1.memcpy((unsigned char*)(out), (unsigned char*)(buff_out), NUM_BLOCK);
+  /*
+  DPCT1010:232: SYCL uses exceptions to report errors and does not use the error codes. The call was replaced with 0. You need to rewrite this code.
+  */
+  //CUDA_CHECK_RETURN(0);
+}
+
+void dequantize(float *code, unsigned char *A, float *out, int n)
+{
+  int num_blocks = n/1024;
+  num_blocks = n % 1024 == 0 ? num_blocks : num_blocks + 1;
+  dpct::device_ext &dev_ct1 = dpct::get_current_device();
+  sycl::queue &q_ct1 = dev_ct1.in_order_queue();
+  sycl::context ctx = q_ct1.get_context();
+  int size = NUM_BLOCK;
+  
+  *((unsigned char **)&buff_A) = sycl::malloc_device(size, A, ctx);

```

Comment:
Is there a good reason to expose the sycl context here, instead of just passing the queue and keeping it hidden?